### ISSUE TITLE
When zooming in/out events are moved across the timeline

### ISSUE THREAD

**Issue by jarnovanrhijn** [namespace-ee/react-calendar-timeline#455]
**Describe the bug**
When zooming in/out events are moved across the timeline and then do not match the start and end time.

**To Reproduce**

1. Go to https://codesandbox.io/s/zr5r289rm4
2. Zoom in / out
3. See events being moved to the left / right

![screen recording 2018-10-24 at 10 01 32 1](https://user-images.githubusercontent.com/2767127/47415731-755ce300-d774-11e8-95f0-9254ab592723.gif)

**Expected behavior**
Events should match the right start- and endtime when zooming in or out.

**Library Version**
React: 16.0.0
RCT: 0.20.0

**Desktop (please complete the following information):**

* OS: MacOS
* Browser Chrome
* Version 69.0.3497.100

**Comment by Ilaiwi** [comment]
@acemac I have the same issue using the sandbox provided by you but it appears when you zoom in and out with the subheader has day units shown (sorry I do not have gif capture on my osx machine)

**Comment by acemac** [comment]
@Ilaiwi Thats no good! I'll get an MacOS machine to test against.

Every 'tick' of my mousewheel triples the amount of timeline shown. And zooming in from any level immediately zooms to a 1 hour visible range. This has always been the behaviour we see on Win10, Win8 and Windows Server.

The video provided there seems to be far more control over the zoom level. Perhaps a race condition with multiple setState called rapidly on the zoom? [Similar to this issue](https://stackoverflow.com/questions/30338577/reactjs-concurrent-setstate-race-condition)

**Comment by Ilaiwi** [comment]
@acemac yes scrolling events seem to have a different behavior in osx than windows. I don't think it is a race condition though because we don't pass state from children to parents but maybe I don't know much about the race condition problem in react much. 

I will debug this after work and comment my findings here

**Comment by acemac** [comment]
@Ilaiwi Good job :) Is this broken? [Example using setState from parent](https://codesandbox.io/s/l52j57yrz9)

Timeline.js line 495

```
    this.setState(
      calculateScrollCanvas(
        visibleTimeStart, 
        visibleTimeEnd, 
        forceUpdateDimensions, 
        items, 
        groups, 
        this.props, 
        this.state))
  }
```

This is a potential race condition if I understand React correctly.

```
    this.setState(prevState => 
      calculateScrollCanvas(
        visibleTimeStart, 
        visibleTimeEnd, 
        forceUpdateDimensions, 
        items, 
        groups, 
        this.props, 
        prevState))
  }
```

**Comment by jarnovanrhijn** [comment]
@acemac & @Ilaiwi, I've tested with your example but the problem still persists. Also tested it in Microsoft Edge (in parallels) and the problem also exists there. If i can help out with testing or debugging let me know.

**Comment by Ilaiwi** [comment]
@acemac  
So the timeline recalculates the items' positions in 4 cases:
1- if you have default time start and default time end on initialization
2- if you change visible time start and visible time end
3- on dragging items and resizing them
4- on mouse wheel and scroll events
all the above cases end up setting state using the function `calculateScrollCanvas` but this function acts smart. if you changed the visible time start and visible time end but did not to cross visible canvas start and visible canvas end then no new calculation will be made.
the above works fine for all the above cases but not for one; when you zoom in and out and you don't cross the boundaries. you can find the code for it [here](https://github.com/namespace-ee/react-calendar-timeline/blob/master/src/lib/utility/calendar.js#L482) 
The fix would be to check if zoom changed if it did then calculate new canvas.
We really really need to write more unit tests for this library

@jarnovanrhijn So I will work on a fix for this and publish it ASAP but for not you can find a fix for it here by adding an `onChangeTime` handler as in the example below.
https://codesandbox.io/s/0oq4y1zy3w

**Comment by acemac** [comment]
@Ilaiwi Thanks for investigating. I will do a PR adding a check to re-draw if the zoom level changes.

**Comment by jarnovanrhijn** [comment]
@Ilaiwi, Thanks for the quick reply and (temp) fix!! Will use the onChangeTime handler for now!
